larpandoracontent/LArReclustering/ThreeDMultiReclusteringAlgorithm.cc

```diff
@@ -0,0 +1,360 @@
+/**
+ *  @file   larpandoracontent/LArReclustering/ThreeDMultiReclusteringAlgorithm.cc
+ *
+ *  @brief  Implementation file for the reclustering algorithm class.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+
+#include "larpandoracontent/LArHelpers/LArPfoHelper.h"
+#include "larpandoracontent/LArHelpers/LArClusterHelper.h"
+
+#include "larpandoracontent/LArReclustering/ThreeDMultiReclusteringAlgorithm.h"
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+ThreeDMultiReclusteringAlgorithm::ThreeDMultiReclusteringAlgorithm() :
+    m_pFomAlgTool {nullptr},
+    m_clusterUListName {"ClustersU"},
+    m_clusterVListName {"ClustersV"},
+    m_clusterWListName {"ClustersW"}
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+ThreeDMultiReclusteringAlgorithm::~ThreeDMultiReclusteringAlgorithm()
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode ThreeDMultiReclusteringAlgorithm::Run()
+{
+    const PfoList *pPfos {nullptr};
+    PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_INITIALIZED, !=,
+                                    PandoraContentApi::GetList(*this, m_pfoListName, pPfos));
+    if (!pPfos || pPfos->empty())
+        return STATUS_CODE_SUCCESS;
+
+    // Ask the FOM alg tool which pfos to recluster
+    PfoList pfosToRecluster;
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, m_pFomAlgTool->GetPfosToRecluster(pPfos, pfosToRecluster));
+    if (pfosToRecluster.empty())
+        return STATUS_CODE_SUCCESS;
+
+    // Remove clusters from the pfos, keeping track of their original pfo
+    std::map<HitType, ClusterList> viewToFreeClusters = { {TPC_3D, {}}, {TPC_VIEW_U, {}}, {TPC_VIEW_V, {}}, {TPC_VIEW_W, {}} };
+    std::map<const Pfo *const, ClusterList> pfoToFreeClusters;
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, FreeClustersFromPfos(pfosToRecluster, viewToFreeClusters, pfoToFreeClusters));
```

Comment:
I see that the only non STATUS_CODE_SUCCESS is given if the cluster cannot be removed from the pfo - i think this suggests that something has gone quite wrong. Should this result in a throw instead i.e. in FreeClustersFromPfo and then the function can be void.

Reply:
To throw or to return an status code is something I don't really have a good grasp on. The `RemovePfo` SDK call returns a `STATUS_CODE_FAILURE` so I am just having this function pass the status code along if that happens. Is there a compelling reason to have this as a throw? Execution gets stopped either way I think/hope

Reply:
Yes, in principle these things behave in the same way (the main difference would be in behaviour of a helper function, where the throw variant doesn't allow a calling function to quiertly/unintentionally ignore a status code), the main reason for the two approaches relates to whether or not the function can/does return a StatusCode. If you don't have that option, the only way to bail the function and flag a problem is with an exception, which is where the THROW_RESULT_IF variant comes in. I think the RETURN variant is fine here, because it isn't ignored by the calling Run function.